hashcat: add page (#5900)
tldr-pages/tldr: pages/linux/hashcat.md

# hashcat

> Fast and advanced password recovery tool.

> > More information: <https://hashcat.net/hashcat/>.

Reviewer: I can't find any useful information regarding the command on this website. What do you think about using <https://manned.org/hashcat>?

Author: fixed

> - Brute-force attack with default hashcat mask:

Reviewer: ```suggestion
- Perform a brute-force attack with default hashcat mask (`hash[:salt]:plain`):
```
Same with the all the other examples except the last.
I'm not sure if the second part of the suggestion is correct, but we should at least somewhere explain what `--attack-mode {{3}}` means.

Author: I do get what you mean with 2 (`hash[:salt]:plain`), could you explain further? 

The default hashcat mask is "?1?2?2?2?2?2?2?3?3?3?3?d?d?d?d" and this works incrementally (from 1 chars to 15). More info can be found here https://hashcat.net/wiki/doku.php?id=hashcat#default_values but I have not found a way to explain this in one line

Author: About the --attack-mode
 # | Mode
 ===+======
  0 | Straight
  1 | Combination
  3 | Brute-force
  6 | Hybrid Wordlist + Mask
  7 | Hybrid Mask + Wordlist

Because the example  explains specifically how to do a  "brute-force attack" (option 3) I think set it as a variable is incorrect, indeed "--attack-mode" is not a "variable" of the burte-force mode attack , instead {{hash_value}} is. What do you think? 

The sentence "Perform a brute-force attack (mode 3) with default hashcat mask" , would work for you?

Reviewer: > I do get what you mean with (`hash[:salt]:plain`), could you explain further?

I didn't know that `mode 3` referred to the attack mode. I thought it was referring to the outfile format mode 3, which is `hash[:salt]:plain`. 

> The sentence "Perform a brute-force attack (mode 3) with default hashcat mask" would work for you?

Sounds good!

> - Perform a brute-force attack (mode 3) with default hashcat mask:

Reviewer: ```suggestion
- Perform a brute-force attack (mode 3) with the default hashcat mask:
```

> `hashcat --hash-type {{id_hash_type}} --attack-mode 3 {{hash}}`

Reviewer: ```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode {{3}} {{hash}}`
```
Same for all others

Author: https://github.com/tldr-pages/tldr/pull/5900#discussion_r627803232

> `hashcat --hash-type {{id_hash_type}} --attack-mode 3 {{hash_value}}`

Reviewer: It's the ID of the hash type, correct? I think `hash_type_id` or just `hash_type` would be clearer.

Author: ok, I prefer to use `hash_type_id` because the parameters expects a number which map an hash type . `hash_type` instead would make you think "MD5" is an accepted value (which is wrong)

Reviewer: Sure

- Perform a brute-force attack (mode 3) with a known pattern of 4 digits:

> `hashcat --hash-type {{id_hash_type}} --attack-mode 3 {{hash_value}} {{"?d?d?d?d"}}`

Reviewer: ```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode 3 {{hash_value}} "{{?d?d?d?d}}"`
```

> - Brute-force attack with max length of 8 chars using the complete charset (numbers + letters + special chars):

Reviewer: ```suggestion
- Brute-force attack with max length of 8 chars using the complete ASCII charset:
```
Is this still correct?

Author: Almost, from the help page:   
   l | abcdefghijklmnopqrstuvwxyz
  u | ABCDEFGHIJKLMNOPQRSTUVWXYZ
  d | 0123456789
  s |  !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
  a | ?l?u?d?s

this mean "?a" cover ASCII from 32 to 126

Author: I replaced it with "all printable ASCII chars"

> - Perform a brute-force attack (mode 3) using 8 (at most) of all printable ASCII chars with:

Reviewer: ```suggestion
- Perform a brute-force attack (mode 3) using at most 8 of all printable ASCII characters:
```

> `hashcat --hash-type {{id_hash_type}} --attack-mode 3 --increment {{hash}} ?a?a?a?a?a?a?a?a`

Reviewer 2: What does `?a?a?a?a?a?a?a?a` mean?

Reviewer: `?a` is a placeholder for any ASCII symbol (<https://manned.org/hashcat#head7>). `?a?a?a?a?a?a?a?a` means there are eight of them.

Reviewer: However it's still kind of a variable so we should use:
```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode 3 --increment {{hash}} {{?a?a?a?a?a?a?a?a}}`
```

Reviewer 2: I think we should also put it in quotes, because most shells interpret `?` as a special character.  For me, `echo ?` doesn't work.  Only `echo "?"`.

Author: Fixed with the brackets but without the quotes because I do not think they are needed in this case. @bl-ue , do you run into any issue running the hashcat command in your shell?

Reviewer: It works without quotes in `bash`, but not in `zsh` or `fish`.

Author: ok, then fixed.

> `hashcat --hash-type {{id_hash_type}} --attack-mode 3 --increment {{hash_value}} {{"?a?a?a?a?a?a?a?a"}}`

Reviewer: ```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode 3 --increment {{hash_value}} "{{?a?a?a?a?a?a?a?a}}"`
```

Author: fixed

- Dictionary attack using the RockYou wordlist of a Kali Linux box:

> `hashcat --hash-type {{id_hash_type}} --attack-mode 0 {{hash}} /usr/share/wordlists/rockyou.txt`

Reviewer: ```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode 0 {{hash}} {{/usr/share/wordlists/rockyou.txt}}`
```
Do all the paths have to be absolute? If not please remove the leading slashes everywhere.

Author: the path is absolute and standard when using Kali linux

Author: i added the brackets

> - Perform a dictionary + rule based attack (mode 0) using the RockYou wordlist mutated with common password variations:

Reviewer: ```suggestion
- Perform a rule based dictionary attack (mode 0) using the RockYou wordlist mutated with common password variations:
```
What do you think about this?

> `hashcat --hash-type {{id_hash_type}} --attack-mode 0 -r {{/usr/share/hashcat/rules/best64.rule}} {{hash_value}} {{/usr/share/wordlists/rockyou.txt}}`

Reviewer: ```suggestion
`hashcat --hash-type {{id_hash_type}} --attack-mode 0 --rules-file={{/usr/share/hashcat/rules/best64.rule}} {{hash_value}} {{/usr/share/wordlists/rockyou.txt}}`
```

Author: fixed (but without '=')

- Perform a combination attack (mode 1) using the concatenation of words from two different custom dictionaries:

`hashcat --hash-type {{id_hash_type}} --attack-mode 1 {{hash_value}} {{/path/to/dictionary1.txt}} {{/path/to/dictionary2.txt}}`

> - Show result of an already-cracked hash:

Reviewer: ```suggestion
- Show result of an already cracked hash:
```
